### ISSUE TITLE
Harmony side effect evaluation doesn’t always work with /*#__PURE__*/

### ISSUE THREAD

**Issue by iamakulov** [webpack/webpack#12472]
# Bug report

**What is the current behavior?**

Webpack has an optimization for [the `sideEffects: false` flag](https://webpack.js.org/configuration/optimization/#optimizationsideeffects). Say, we have a module `foo` with `"sideEffects": false` in its `package.json`. If we import this module but don’t use any of module’s exports, webpack will remove this module from the bundle completely.

I’ve stumbled upon a case where this doesn’t work as expected. Specifically, if we import a function from the `foo` module, call it, but prefix the call with `/*#__PURE__*/`, the module is retained in the bundle – even though Terser removes the call:

```js
// This works:
import { someFunc } from 'foo';
// → someFunc is unused – so it’s removed, and 'foo' is dropped from the bundle


// This doesn’t:
import { someFunc } from 'foo';
/*#__PURE__*/ someFunc();

// → someFunc’s result is unused – so it’s removed. However, 'foo' is kept in the bundle
// (along with its top-level code)
```

**If the current behavior is a bug, please provide the steps to reproduce.**

GitHub repository: https://github.com/iamakulov/webpack-harmony-pure-repro

1. Run `yarn` to install dependencies
2. Run `yarn webpack --config webpack.config.js` to build the repo
3. Open `dist/main.js`

**What is the expected behavior?**

If the side effect evaluation was performed properly, I’d expect `dist/main.js` to look like this:

```js
(()=>{"use strict";console.log("running root")})();
```

(You can get this result by commenting out line 5 in `index.js`.)

**What is the actual behavior?**

`dist/main.js` looks like this:

```js
(()=>{"use strict";console.log("running child-module"),console.log("running root")})();
```

**Other relevant information:**
webpack version: 5.16.0
Node.js version: 12.20.1
Operating System: macOS Big Sur (Intel)
Additional tools: -

**Comment by sokra** [comment]
> ```js
> /*#__PURE__*/ someFunc();
> ```

Per definition this line has no side effect and the result is never used. It would always be removed. Why are you writing this line at all? Did you write it by hand or is this the output of some other tool?

**Comment by iamakulov** [comment]
> Per definition this line has no side effect and the result is never used. It would always be removed. Why are you writing this line at all? Did you write it by hand or is this the output of some other tool?

In my case, this was a manual optimization done for a private component library.

* The library was distributed as multiple packages:

   ```
   lib
   lib-fonts
   lib-icons
   lib-utils
   ```

* Each package had a single (Rollup-produced) bundle
* Packages were consumed by webpack-based applications
* The `lib` package had a reexport from `lib-fonts` which wasn’t properly tree-shaken out – which is why we tried adding `/*#__PURE__#/`

**Comment by alexander-akait** [comment]
bump

**Comment by vankop** [comment]
```
/*#__PURE__*/ someFunc();
```
is weird.. (what is the reason to call pure function and doesn't use a result)  maybe in issue example it is useful, but in general can lead to aggressive tree-shaking.

I think this should not be implemented..

**Comment by iamakulov** [comment]
Gah, actually, React is a bad example because it doesn’t have `sideEffects: false` in the bundle.

`styled-components` is a better one: they do have `sideEffects: [...]` in `package.json`, and they also use `babel-plugin-styled-components` to insert `/*#__PURE__*/` in front of style rules:

```js
import { css } from 'styled-components'

const buttonStyles = css`color: red`
```

↓

```js
import { css } from 'styled-components'

const buttonStyles = /*#__PURE__*/ css`color: red`
```

Updated the repro once again.

**Comment by vankop** [comment]
yes, original issue doesn't work since there is no assignment ( in this case webpack doesn't check PURE directive ). As I described this looks weird when result of pure function doesn't used.. Supporting this could lead to aggressive tree-shaking..
________

To fix original issue you can do
```js
import { someFunc } from 'foo';
const a = /*#__PURE__*/ someFunc();
```

**Comment by iamakulov** [comment]
@vankop Oh yeah, thank you! Can confirm 

```js
const a = /*#__PURE__*/ someFunc();
```

works as intended (the `foo` module isn’t bundled anymore). Happy to keep this issue closed 👍
